terrain/terrain.cc

```diff
@@ -0,0 +1,623 @@
+#include <math.h>
+#include <vector>
+#include <map>
+#include <string>
+#include <algorithm>
+
+#include "terrain.h"
+#include "perlin.h"
+#include "mc.h"
+#include "../vector.h"
+#include "../noiser.h"
+
+namespace Terrain {
+
+float isoLevel = 0.0;
+
+float noiseScale = 3.0;
+int octaves = 8;
+float persistence = 1.15;
+float lacunarity = 1.6;
+float floorOffset = 20.0;
+float hardFloor = 2.0;
+float hardFloorWeight = 3.06;
+float noiseWeight = 6.09;
+
+int NUM_BUFFER = 8;
+
+float* generateTerrain(
+    float chunkSize, int chunkCount, int segment, int vertexBufferSizeParam, int faceBufferSizeParam
+) {
+
+    // std::vector<Vector3> vertices = {};
+    // std::vector<int> indices = {};
+
+    int totalChunkCount = pow(chunkCount, 3);
+
+    int cellCount = totalChunkCount * segment * segment;
+
+    int maxVertexCount = cellCount * vertexBufferSizeParam;
+    int maxIndexCount = cellCount * faceBufferSizeParam;
+
+
+    float *vertexBuffer = (float*)malloc(maxVertexCount * 3 * sizeof(float));
```

Comment:
What is the point of this? Below you will copy into this from a `std::vector`, and you know the exact size.

So this is not saving any computation, it just makes the user pass extra arguments, it wastes memory, and if you guess the wrong arguments it explodes.

We can delete these two arguments: `int vertexBufferSizeParam, int faceBufferSizeParam` and allocate the exact size.